### ISSUE TITLE
signs do not refresh after git add and git commit

### ISSUE THREAD

**Issue by jdhao** [lewis6991/gitsigns.nvim#103]
**Describe the bug**

I have noticed that signs do not refresh automatically after I have git add and commit the current file. I have to manually `:e %` to make the diff signs disappear.

**To Reproduce**

init.vim:

```vim
set runtimepath+=/home/haojiedong/.local/share/nvim/plugged/gitsigns.nvim
set runtimepath+=/home/haojiedong/.local/share/nvim/plugged/plenary.nvim

lua << EOF
require('gitsigns').setup {
  signs = {
    add          = {text = '+'},
    change       = {text = '~'},
    delete       = {text = '_', show_count=true},
    topdelete    = {text = '‾', show_count=true},
    changedelete = {text = '~', show_count=true},
  },
}
EOF
```

1. Go to a git project and open a file with the above config, `nvim -u init.vim some_file`
2. Make changes to this file, then `git add some_file` and `git commit` it.
3. Notice that git signs is still in `some_file`.

only after `:e %`, the git signs disappear.

**Additional context**

System: Linux
Neovim version: `v0.5.0-dev+1133-g0869cbd55`

**Comment by lewis6991** [comment]
I'm not able to reproduce this.

Can you try adding `debug_mode=true` to your config and after you run `git add` on your file run: 
```
:lua require'gitsigns'.debug_messages()
```

And paste the output here. You should see something like:

```sh
attach(1): Attaching
dprint(nil): Running: git rev-parse --show-toplevel --absolute-git-dir --abbrev-ref HEAD
dprint(nil): Running: git --no-pager ls-files --stage --others --exclude-standard /path/to/some_file
watch_index(1): Watching index

dprint(nil): Running: git --no-pager show :0:some_file
update(1): updates: 1, jobs: 3
dprint(nil): Running: git --no-pager show :0:some_file
update(1): updates: 2, jobs: 4
dprint(nil): Running: git --no-pager show :0:some_file
update(1): updates: 3, jobs: 5
# These messages happen every time you edit the buffer

attach(1): Already attached

watcher_cb(1): Index update  
# this should trigger on any git command that updates the index (git add, git commit, etc)

dprint(nil): Running: git rev-parse --show-toplevel --absolute-git-dir --abbrev-ref HEAD
dprint(nil): Running: git --no-pager ls-files --stage --others --exclude-standard /path/to/some_file
dprint(nil): Running: git --no-pager show :0:some_file
update(1): updates: 4, jobs: 8
```

You should see `watcher_cb(1): Index update` which happens when a change is detected in `.git/index`

**Comment by jdhao** [comment]
okay, I have followed you instructions, the new config:

```
set runtimepath+=/home/haojiedong/.local/share/nvim/plugged/gitsigns.nvim
set runtimepath+=/home/haojiedong/.local/share/nvim/plugged/plenary.nvim

lua << EOF
require('gitsigns').setup {
  signs = {
    add          = {text = '+'},
    change       = {text = '~'},
    delete       = {text = '_', show_count=true},
    topdelete    = {text = '‾', show_count=true},
    changedelete = {text = '~', show_count=true},
  },
  debug_mode = true
}
EOF
```

The debug message I got when run `:lua require'gitsigns'.debug_messages()`

```
dprint(nil): Deriving GitSignsChange from DiffChange
dprint(nil): Deriving GitSignsDelete from DiffDelete
dprint(nil): Deriving GitSignsAdd from DiffAdd
attach(1): Attaching
dprint(nil): Running: git rev-parse --show-toplevel --absolute-git-dir --abbrev-ref HEAD
dprint(nil): Running: git --no-pager ls-files --stage --others --exclude-standard  /path/to/my_file
watch_index(1): Watching index
dprint(nil): Running: git --no-pager show :0: path/to/my_file
watcher_cb(1): Index update
dprint(nil): Running: git rev-parse --show-toplevel --absolute-git-dir --abbrev-ref HEAD
dprint(nil): Running: git --no-pager ls-files --stage --others --exclude-standard path/to/my_file
update(1): updates: 1, jobs: 5
watcher_cb(1): File not changed
```

I did the see the `Index update` message, but the diff sign on the sign column
does not disappear. If I open another file and use `:b my_file`, I can see that
the diff sign disappear.

Note that I have used a similar plugin, vim-signify. With it, as soon as I git
add a file, the diff signs immediately disappear. I do not think it is a nvim
issue.

**Comment by lewis6991** [comment]
Thanks for doing that.

As the log indicates, gitsigns thinks the file hasn't changed in the index update so hasn't bothered to update the signs.

Can you run:
```sh
git --no-pager ls-files --stage --others --exclude-standard path/to/my_file
```
Before and after you run `git-add`? It should return something like:

```
100644 29931f323ca45c803d254a360e926eaf33b3fd51 0      /path/to/file
```
However the hash should be different in each output. If the hash is the same then that means `git-add` didn't make any changes to the index for `/path/to/file`.

> Note that I have used a similar plugin, vim-signify. With it, as soon as I git
add a file, the diff signs immediately disappear. I do not think it is a nvim
issue.

While vim-signify does roughly the same thing as gitsigns, its implementation is completely different. They both use: different languages (lua vs viml), different API's and even a different method for updating signs. It's not really useful to compare the two other than how good the UX is. If there is an issue with say `vim.loop.new_fs_poll()` or `nvim_buf_attach()` (signify doesn't use these) then that would be a neovim issue. However since we are seeing the index update happening, it looks like a potential issue in `gitsigns.git.file_info()`.

**Comment by jdhao** [comment]
Output of command  `git --no-pager ls-files --stage --others --exclude-standard path/to/my_file`:

+ before git add: `100644 1777c69598a55fbc21b86a6089d0baad0bc3e124 0       test.md`
+ after git add: `100644 353f7c0b619ad5faf230c76e256a871dd9a819ef 0       test.md`

Output of `:lua require'gitsigns'.dump_cache()`:

+ before git add:

```
{ {
    abbrev_head = "master",
    file = "/home/haojiedong/projects/play_ground/git_test/test.md",
    gitdir = "--absolute-git-dir",
    has_conflicts = false,
    hunks = { {
        added = {
          count = 2,
          start = 2
        },
        dend = 3,
        head = "@@ -1 +2,2 @@",
        lines = { "+", "+modification" },
        removed = {
          count = 0,
          start = 1
        },
        start = 2,
        type = "add"
      } },
    index_watcher = <userdata 1>,
    mode_bits = "100644",
    object_name = "1777c69598a55fbc21b86a6089d0baad0bc3e124",
    pending_signs = {},
    relpath = "test.md",
    staged = "/tmp/lua_tXx1Zo",
    staged_diffs = {},
    staged_text = { "some test" },
    toplevel = "/home/haojiedong/projects/play_ground/git_test"
  } }
```
+ after git add:

```
{ {
    abbrev_head = "master",
    file = "/home/haojiedong/projects/play_ground/git_test/test.md",
    gitdir = "--absolute-git-dir",
    has_conflicts = false,
    hunks = { {
        added = {
          count = 2,
          start = 2
        },
        dend = 3,
        head = "@@ -1 +2,2 @@",
        lines = { "+", "+modification" },
        removed = {
          count = 0,
          start = 1
        },
        start = 2,
        type = "add"
      } },
    index_watcher = <userdata 1>,
    mode_bits = "100644",
    object_name = "1777c69598a55fbc21b86a6089d0baad0bc3e124",
    pending_signs = {},
    relpath = "test.md",
    staged = "/tmp/lua_tXx1Zo",
    staged_diffs = {},
    staged_text = { "some test" },
    toplevel = "/home/haojiedong/projects/play_ground/git_test"
  } }
```

**Comment by jdhao** [comment]
Output of command  `git --no-pager ls-files --stage --others --exclude-standard path/to/my_file`:

+ before git add: `100644 1777c69598a55fbc21b86a6089d0baad0bc3e124 0       test.md`
+ after git add: `100644 353f7c0b619ad5faf230c76e256a871dd9a819ef 0       test.md`

Output of `:lua require'gitsigns'.dump_cache()`:

+ before git add:

```
{ {
    abbrev_head = "master",
    file = "/home/haojiedong/projects/play_ground/git_test/test.md",
    gitdir = "--absolute-git-dir",
    has_conflicts = false,
    hunks = { {
        added = {
          count = 2,
          start = 2
        },
        dend = 3,
        head = "@@ -1 +2,2 @@",
        lines = { "+", "+modification" },
        removed = {
          count = 0,
          start = 1
        },
        start = 2,
        type = "add"
      } },
    index_watcher = <userdata 1>,
    mode_bits = "100644",
    object_name = "1777c69598a55fbc21b86a6089d0baad0bc3e124",
    pending_signs = {},
    relpath = "test.md",
    staged = "/tmp/lua_tXx1Zo",
    staged_diffs = {},
    staged_text = { "some test" },
    toplevel = "/home/haojiedong/projects/play_ground/git_test"
  } }
```
+ after git add:

```
{ {
    abbrev_head = "master",
    file = "/home/haojiedong/projects/play_ground/git_test/test.md",
    gitdir = "--absolute-git-dir",
    has_conflicts = false,
    hunks = { {
        added = {
          count = 2,
          start = 2
        },
        dend = 3,
        head = "@@ -1 +2,2 @@",
        lines = { "+", "+modification" },
        removed = {
          count = 0,
          start = 1
        },
        start = 2,
        type = "add"
      } },
    index_watcher = <userdata 1>,
    mode_bits = "100644",
    object_name = "1777c69598a55fbc21b86a6089d0baad0bc3e124",
    pending_signs = {},
    relpath = "test.md",
    staged = "/tmp/lua_tXx1Zo",
    staged_diffs = {},
    staged_text = { "some test" },
    toplevel = "/home/haojiedong/projects/play_ground/git_test"
  } }
```

**Comment by lewis6991** [comment]
Ok, I think we are slowly narrowing down the issue. From the information given, we know that:
- The index handler is triggering
- The cache is being populated properly
- The object hash **is** changing in the git index when you run `git add`.

From this I strongly suspect this is a latency issue. It appears that the index handler is triggering (and completing) before the object database has been updated. There are a couple things we can try here:

1. After running `git add`, run another command that updates `.git/index`. Either add another file or maybe `touch .git/index` will do it.
2. Try increasing the default value of `config.watch_index.interval`. The current default is 1000ms; maybe try setting it to something like `2000` or higher and see if that fixes things.
3. If that doesn't work trying injecting a delay in `index_update_handler()` in `lua/gitsigns.lua` (around line 430). Add `vim.loop.sleep(2000)` just before the call to `await(git.get_repo_info, ...)`.

If any of these work, then that probably confirms this as a latency issue.

**Comment by jdhao** [comment]
1. git add another file, after `git add test.md`: does not work
2.  increase `watch_index.interval` to 3000, 4000 and 5000: does not work
3.  Add `vim.loop.sleep(2000)` before `await(git.get_reop_info. bcache.toplevel)`: does not work either.

**Comment by lewis6991** [comment]
Not a latency issue then...

I'm pretty much out of ideas at this point. Without being able to recreate the issue, I'm not sure how much help I can provide. We have tests in the testsuite for this so I'm sure the root issue is probably quite nuanced and specific to your setup, otherwise this issue would have been raised already by a numerous people.

Final tips:
- Make sure `index_update_handler()` is triggering because of your `git-add` and not for another reason. Another guess is that the handler is being called immeditely with an error. You should be able to change these lines:
  ```lua
  local function index_update_handler(cbuf)
     return sync(function()
       ...
  ```

  Into:
  ```lua
  local function index_update_handler(cbuf)
     return sync(function(err, prev, cur)
      print(vim.inspect(err))
      print(vim.inspect(prev))
      print(vim.inspect(cur))
       ...
  ```

  Also see: [fs_poll_start](https://github.com/luvit/luv/blob/master/docs.md#uvfs_poll_startfs_poll-path-interval-callback)

- Add lots of print statements in places to see what's going on. Specifically in `git.file_info()` and `update_index_handler()`

Please let me know if you can narrow this down any further. I'd really like to fix this issue if we can figure out the root cause.

**Comment by jdhao** [comment]
Sorry, I am a complete newbie for lua, though I am mildly procient in vimscript. Okay, I have added the three print message as instrcuted. 

After git add, I am seeing the following messages:

```
"ENOENT"                            
{                                   
  atime = {                         
    nsec = 0,                       
    sec = 0                         
  },                                
  birthtime = {                     
    nsec = 0,                       
    sec = 0                         
  },                                
  blksize = 0,                      
  blocks = 0,                       
  ctime = {                         
    nsec = 0,                       
    sec = 0                         
  },                                
  dev = 0,                          
  flags = 0,                        
  gen = 0,                          
  gid = 0,                          
  ino = 0,                          
  mode = 0,                         
  mtime = {                         
    nsec = 0,                       
    sec = 0                         
  },                                
  nlink = 0,                        
  rdev = 0,                         
  size = 0,                         
  uid = 0                           
}                                   
{                                   
  atime = {                         
    nsec = 0,                       
    sec = 0                         
  },                                
  birthtime = {                     
    nsec = 0,                       
    sec = 0                         
  },                                
  blksize = 0,                      
  blocks = 0,                       
  ctime = {                         
    nsec = 0,                       
    sec = 0                         
  },                                
  dev = 0,                          
  flags = 0,                        
  gen = 0,                          
  gid = 0,                          
  ino = 0,                          
  mode = 0,                         
  mtime = {                         
    nsec = 0,                       
    sec = 0                         
  },                                
  nlink = 0,                        
  rdev = 0,                         
  size = 0,                         
  uid = 0                           
}                                   
```

The `ENOENT` should be the err message. I looked up it a bit, find that it may mean `no such file or directory`. Is this the culprint?

**Comment by lewis6991** [comment]
There it is!

Are you sure this is triggering after `git-add` and not before?

In `watch_index()` add `print(vim.inspect(index))` and check that path exists.

Basically the issue is the index watcher is watchning a non-existent path. Solution should be as simple as fixing the path we need to watch.

**Comment by jdhao** [comment]
Okay, I `git reset` , restart nvim and open the file again, it seems the above message still persists. So it exists before I git add.

I add `print(vim.inspect(index))` in function `watch_index()`, what I see is `"--absolute-git-dir/index" `.  I do not think it is a valid index?

**Comment by lewis6991** [comment]
What do you get when you run:

```sh
git rev-parse --show-toplevel  --absolute-git-dir --abbrev-ref HEAD
```

**Comment by jdhao** [comment]
THe output of command `git rev-parse --show-toplevel  --absolute-git-dir --abbrev-ref HEAD`:

```
/home/haojiedong/projects/play_ground/git_test
--absolute-git-dir
master
```

**Comment by lewis6991** [comment]
What's your git version?

**Comment by jdhao** [comment]
I am using git 1.8.3.1 on a CentOS 7.4 server.

**Comment by lewis6991** [comment]
Ok that is a pretty ancient version. Might be able to get this to work though, what's the output of:
```sh
git rev-parse --show-toplevel --git-dir --abbrev-ref HEAD
```

**Comment by jdhao** [comment]
THe output of `git rev-parse --show-toplevel --git-dir --abbrev-ref HEAD`:

```
/home/haojiedong/projects/play_ground/git_test
.git
master
```

**Comment by lewis6991** [comment]
Ok I'll get back to you whether we can get this to work. However I expect you'll hit other issues because the version of git you're using is quite old.

Any chance you can upgrade your version of git to something more recent. `--absolute-git-dir` was added in 2017.

**Comment by jdhao** [comment]
I do wish to use a newer version of git. However, since I am not root, it is quite difficult for me to upgrade the git version.

**Comment by lewis6991** [comment]
Ok no worries let me see what I can do.

**Comment by lewis6991** [comment]
Checkout the latest updates. Does it improve things at all?

**Comment by jdhao** [comment]
I updated to the latest version. As soon as I open a file, I am see the following error:

> Error executing luv callback:                                                                                                          
/root/gitsigns.nvim/lua/gitsigns/async.lua:56: /root/gitsigns.nvim/lua/gitsigns.lua:422: Error detected in index watcher for buffer 1:
ENOENT

**Comment by lewis6991** [comment]
What does the cache now say?

**Comment by jdhao** [comment]
The output of `lua require'gitsigns'.dump_cache()`

```
{ {
    abbrev_head = "HEAD",
    file = "/root/git_test/test.md",
    gitdir = "/root/git_test/.git",
    has_conflicts = false,
    hunks = { {
        added = {
          count = 1,
          start = 1
        },
        dend = 1,
        head = "@@ -0 +1,1 @@",
        lines = { "+first line" },
        removed = {
          count = 0,
          start = 0
        },
        start = 1,
        type = "add"
      } },
    index_watcher = <userdata 1>,
    pending_signs = {},
    relpath = "test.md",
    staged = "/tmp/lua_vnodYk",
    staged_diffs = {},
    staged_text = {},
    toplevel = "/root/git_test"
  } }
```

**Comment by lewis6991** [comment]
Ok now we have `gitdir = "/root/git_test/.git"` which is better than before. Does `/root/git_test/.git/index` exist as a file?

**Comment by jdhao** [comment]
No, the file /root/git_test/.git/index does not exist.

**Comment by lewis6991** [comment]
...ok so what does exist? (Contents of `.git`)?

Help me out here.

**Comment by jdhao** [comment]
Content of `.git` directory:

> branches/  config  description  HEAD  hooks/  info/  objects/  refs/



> No, the file /root/git_test/.git/index does not exist.
> 
> Okay, this repository has been freshly setup. I haven't done any git add yet. I tried to git add test.md.
> 
> Then make changes to test.md, after that, gitsign.nvim seems to work fine.

As I have said, it seems that I need to git add one time, before gitsigns.nvim can work normally.

**Comment by lewis6991** [comment]
Ah ok, I'll consider that a separate issue (index watcher on fresh repos).

Otherwise is the issue now resolved?

**Comment by jdhao** [comment]
yeah. I consider it solved. cheers 😃 👍

**Comment by lewis6991** [comment]
🎉

**Comment by jdhao** [comment]
So should I create another issue or you will do it yourself?

**Comment by lewis6991** [comment]
Would appreciate if you could, kinda of trying to fix the other reported bug. Otherwise I'll raise it later.

**Comment by jdhao** [comment]
Okay, no worries. I will create a new one.
